File: app/java/src/main/java/io/infralovers/nomadvaultmysql/controller/AppController.java
package io.infralovers.nomadvaultmysql.controller;

import io.infralovers.nomadvaultmysql.model.Customer;
import io.infralovers.nomadvaultmysql.model.CustomerInput;
import io.infralovers.nomadvaultmysql.service.CustomerRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@Controller
public class AppController {

    private static final Logger log = LoggerFactory.getLogger(AppController.class);

    private final CustomerRepository repository;

    public AppController(CustomerRepository repository) {
        this.repository = repository;
    }

    // ── Health / API ──────────────────────────────────────────────────────────

    @GetMapping("/health")
    @ResponseBody
    public ResponseEntity<String> health() {
        if (!repository.isInitialized()) {
            return ResponseEntity.status(500).body("Unhealthy - no database");
        }
        return ResponseEntity.ok("Healthy");
    }

    @GetMapping("/customers")
    @ResponseBody
    public List<Customer> getCustomers() throws Exception {
        return repository.getCustomerRecords(50, false);
    }

    @GetMapping("/customer")
    @ResponseBody
    public ResponseEntity<?> getCustomer(@RequestParam(name = "cust_no", required = false) Integer custNo) throws Exception {
        if (custNo == null) {
            return ResponseEntity.status(500)
                    .body("<html><body>Error: cust_no is a required argument for the customer endpoint.</body></html>");
        }
        return ResponseEntity.ok(repository.getCustomerRecord(custNo));
    }

    @PostMapping("/customers")
    @ResponseBody
    public List<Customer> createCustomer(CustomerInput input) throws Exception {
        return repository.insertCustomerRecord(input);
    }

    @PutMapping("/customers")
    @ResponseBody
    public List<Customer> updateCustomer(CustomerInput input) throws Exception {
        return repository.updateCustomerRecord(input);
    }

    // ── UI ────────────────────────────────────────────────────────────────────

    @GetMapping("/")
    public String index() {
        return "index";
    }

    @GetMapping("/records")
    public String records(Model model) throws Exception {
        model.addAttribute("results", repository.getCustomerRecords(50, false));
        return "records";
    }

    @GetMapping("/dbview")
    public String dbview(Model model) throws Exception {
        model.addAttribute("results", repository.getCustomerRecords(50, true));
        return "dbview";
    }

    @GetMapping("/add")
    public String addForm() {
        return "add";
    }

    @PostMapping("/add")
    public String addSubmit(CustomerInput input, Model model) throws Exception {
        List<Customer> records = repository.insertCustomerRecord(input);
        model.addAttribute("results", records);
        model.addAttribute("recordAdded", true);
        return "records";
    }

    @GetMapping("/update")
    public String updateForm(@RequestParam(name = "cust_no", required = false) Integer custNo, Model model) throws Exception {
        if (custNo != null) {
            List<Customer> found = repository.getCustomerRecord(custNo);
            if (!found.isEmpty()) {
                model.addAttribute("customer", found.get(0));
            }
        }
        return "update";
    }

    @PostMapping("/update")
    public String updateSubmit(CustomerInput input, Model model) throws Exception {
        List<Customer> records = repository.updateCustomerRecord(input);
        model.addAttribute("results", records);
        model.addAttribute("recordUpdated", true);
        return "records";
    }
}

File: app/dotnet/Controllers/AppController.cs
using Microsoft.AspNetCore.Mvc;
using NomadVaultMySqlDotnet.Models;
using NomadVaultMySqlDotnet.Services;

namespace NomadVaultMySqlDotnet.Controllers;

[IgnoreAntiforgeryToken]
public sealed class AppController : Controller
{
    private readonly CustomerRepository _repository;

    public AppController(CustomerRepository repository)
    {
        _repository = repository;
    }

    [HttpGet("/health")]
    public IActionResult Health()
    {
        if (!_repository.IsInitialized)
        {
            return StatusCode(500, "Unhealthy - no database");
        }

        return Content("Healthy");
    }

    [HttpGet("/customers")]
    public async Task<IActionResult> GetCustomers(CancellationToken cancellationToken)
    {
        var customers = await _repository.GetCustomerRecordsAsync(cancellationToken: cancellationToken);
        return Json(customers.Select(ToApiShape));
    }

    [HttpGet("/customer")]
    public async Task<IActionResult> GetCustomer([FromQuery(Name = "cust_no")] int? customerNumber, CancellationToken cancellationToken)
    {
        if (!customerNumber.HasValue)
        {
            return StatusCode(500, "<html><body>Error: cust_no is a required argument for the customer endpoint.</body></html>");
        }

        var customers = await _repository.GetCustomerRecordAsync(customerNumber.Value, cancellationToken);
        return Json(customers.Select(ToApiShape));
    }

    [HttpPost("/customers")]
    public async Task<IActionResult> CreateCustomer([FromForm] CustomerInput customer, CancellationToken cancellationToken)
    {
        var records = await _repository.InsertCustomerRecordAsync(customer, cancellationToken);
        return Json(records.Select(ToApiShape));
    }

    [HttpPut("/customers")]
    public async Task<IActionResult> UpdateCustomer([FromForm] CustomerInput customer, CancellationToken cancellationToken)
    {
        var records = await _repository.UpdateCustomerRecordAsync(customer, cancellationToken);
        return Json(records.Select(ToApiShape));
    }

    [HttpGet("/")]
    public IActionResult Index()
    {
        return View("Index");
    }

    [HttpGet("/records")]
    public async Task<IActionResult> Records(CancellationToken cancellationToken)
    {
        var records = await _repository.GetCustomerRecordsAsync(cancellationToken: cancellationToken);
        return View("Records", new RecordsPageModel { Results = records });
    }

    [HttpGet("/dbview")]
    public async Task<IActionResult> DbView(CancellationToken cancellationToken)
    {
        var records = await _repository.GetCustomerRecordsAsync(raw: true, cancellationToken: cancellationToken);
        return View("DbView", new RecordsPageModel { Results = records });
    }

    [HttpGet("/add")]
    public IActionResult Add()
    {
        return View("Add");
    }

    [HttpPost("/add")]
    public async Task<IActionResult> AddSubmit([FromForm] CustomerInput customer, CancellationToken cancellationToken)
    {
        var records = await _repository.InsertCustomerRecordAsync(customer, cancellationToken);
        return View("Records", new RecordsPageModel { Results = records, RecordAdded = true });
    }

    [HttpGet("/update")]
    public async Task<IActionResult> Update([FromQuery(Name = "cust_no")] int? customerNumber, CancellationToken cancellationToken)
    {
        if (customerNumber.HasValue)
        {
            var records = await _repository.GetCustomerRecordAsync(customerNumber.Value, cancellationToken);
            var existing = records.FirstOrDefault();
            if (existing is not null)
            {
                return View("Update", existing);
            }
        }

        return View("Update", (CustomerRecord?)null);
    }

    [HttpPost("/update")]
    public async Task<IActionResult> UpdateSubmit([FromForm] CustomerInput customer, CancellationToken cancellationToken)
    {
        var records = await _repository.UpdateCustomerRecordAsync(customer, cancellationToken);
        return View("Records", new RecordsPageModel { Results = records, RecordUpdated = true });
    }

    private static object ToApiShape(CustomerRecord record)
    {
        return new
        {
            customer_number = record.CustomerNumber,
            birth_date = record.BirthDate,
            first_name = record.FirstName,
            last_name = record.LastName,
            create_date = record.CreateDate,
            ssn = record.Ssn,
            ccn = record.Ccn,
            address = record.Address,
            salary = record.Salary
        };
    }
}
